pkg/config/configfile.go

```diff
@@ -0,0 +1,369 @@
+	if len(cf.watchers) == 1 {
+		cf.runWatcher()
+	}
+
+	return h.id
+}
+
+// RemoveChangeHandler removes the change handler with the provided identifier if it exists. True
+// is returned if the handler was removed (it existed), false otherwise.
+func (cf *ConfigFile) RemoveChangeHandler(id HandlerID) bool {
+	cf.watchLock.Lock()
+	defer cf.watchLock.Unlock()
+
+	for i := range cf.watchers {
+		if cf.watchers[i] != nil && cf.watchers[i].id == id {
+			copy(cf.watchers[i:], cf.watchers[i+1:])
+			cf.watchers[len(cf.watchers)-1] = nil
+			cf.watchers = cf.watchers[:len(cf.watchers)-1]
+
+			// stop watching the file for changes if there are no more external watchers
+			if len(cf.watchers) == 0 {
+				cf.stopWatcher()
+			}
+
+			return true
+		}
+	}
+	return false
+}
+
+// RemoveAllHandlers removes all added handlers
+func (cf *ConfigFile) RemoveAllHandlers() {
+	cf.watchLock.Lock()
+	defer cf.watchLock.Unlock()
+
+	cf.watchers = nil
+
+	cf.stopWatcher()
+}
+
+// runWatcher creates a go routine which will poll the stat of a storage target on a specific
+// interval and dispatch created, modified, and deleted events for that file.
+func (cf *ConfigFile) runWatcher() {
+	if cf.watchStop != nil {
+		log.Warningf("Run watcher already running for file: %s", cf.file)
+		return
+	}
+
+	cf.watchStop = make(chan struct{})
+
+	go func() {
+		first := true
+
+		var last time.Time
+		var exists bool
+
+		for {
+			// Each iteration, check for the stop trigger, or wait 10 seconds
+			select {
+			case <-cf.watchStop:
+				return
+			case <-time.After(10 * time.Second):
+			}
+
+			// Query stat on the file, on errors other than exists,
+			// we'll need to log the error, and perhaps limit the retries
+			st, err := cf.Stat()
+			if err != nil && !os.IsNotExist(err) {
+				log.Errorf("Storage Stat Error: %s", err)
+				continue
+			}
+
+			// On first iteration, set exists and last modification time (if applicable)
+			// and flip flag
+			if first {
+				exists = !os.IsNotExist(err)
+				if exists {
+					last = st.ModTime
+				}
+				first = false
+				continue
+			}
+
+			// File does not exist in storage, need to check to see if that is different
+			// from last state check
+			if os.IsNotExist(err) {
+				// check to see if the file has gone from exists to !exists
+				if exists {
+					exists = false
+					cf.onFileChange(ChangeTypeDeleted, nil)
+				}
+				continue
+			}
+
+			// check to see if the file has gone from !exists to exists
+			if !exists {
+				data, err := cf.internalRead(true)
+				if err != nil {
+					log.Warningf("Read() Error: %s\n", err)
+					continue
+				}
+				exists = true
+				last = st.ModTime
+				cf.onFileChange(ChangeTypeCreated, data)
+				continue
+			}
+
+			mtime := st.ModTime
+			if mtime != last {
+				last = mtime
+				data, err := cf.internalRead(true)
+				if err != nil {
+					log.Errorf("Read() Error: %s\n", err)
+					continue
+				}
+				cf.onFileChange(ChangeTypeModified, data)
+			}
+		}
+	}()
```

Comment:
I'm having trouble following this. I think it could use better comments and/or restructuring in code. I don't have your wizard power :mage_man:.

Reply:
Might be hard to say, but can you elaborate, Michael? Where do you get lost? Is there anywhere, in particular, that could use some attention?

Reply:
The parts I find hardest to follow involve repeated use of `first`, `last`, and `exists` in multiple blocks. Making the variable names a little more descriptive would help.

I think another thing that's tripping me up is that this feels like fairly complex state management and having it spelled out more exactly in comments would do a lot to help future me if I ever have to make modifications; I would feel really uncomfortable making modifications to this code as it stands now. Still trying to put my finger on exactly what makes this difficult -- another thing that could improve my confidence would be unit tests, though I don't know how feasible those are here.

Reply:
There are some very obvious code comments which describe first, last, and exists in the code when each is used. 

`first` and `last`
```
// On first iteration, set exists and last modification time (if applicable)
// and flip flag
```

`exists`
```
// File does not exist in storage, need to check to see if that is different
// from last state check
if os.IsNotExist(err) {
	// check to see if the file has gone from exists to !exists
	if exists {
```

There is a first iteration flag which basically signals this is the first time we're looping and the `last` modification time is unset, so set it. For a file to have a last modification time, it has to exist. The change monitor will notify the handler on creation, on change, and on removal. To test change, you need to know what the time was before (last) compared to the current state. A file is new when it goes from !exists to exists, and deletion when it goes from exists to !exists.

Reply:
I think I was tired when I first read through this. I've just re-read the whole thing and it makes perfect sense. Sorry!